### PR TITLE
Implemented Backend for Currency Handling and Frontend Integration

### PR DESCRIPTION
**Pull request by dro42**
### **Pull Request Title**

[PR] Implemented Backend for Currency Handling and Frontend Integration

---

### **Description**

This pull request introduces the following changes:
- Backend implementation for currency management (handled by Beni).
- Frontend updates for currency functionality and user interaction (handled by Andreas).
- Added logging and comprehensive tests to ensure seamless functionality and compatibility.

---

### **Related Issue**

Closes #<Issue-ID>

---

### **Changes Made**

- [x] Implemented backend logic for managing currency.
- [x] Designed and integrated the frontend UI for currency selection and display.
- [x] Added logging for key events and operations related to currency handling.
- [x] Created tests to verify backend and frontend functionality.

---

### **Testing**

1. Unit tests:
   - Created and executed tests for backend currency logic.
   - Verified frontend integration through automated test cases.
   
2. Code coverage:
   - Ensured that all new code meets or exceeds the required code coverage thresholds.

3. Manual testing:
   - Conducted manual testing for currency selection, logging functionality, and edge cases.

---

### **Checklist**

- [x] Code adheres to the project’s style guidelines.
- [x] Comments and documentation added as needed.
- [x] Related issue linked (see above).
- [x] Tests written and passing.
- [x] No conflicts with the base branch.

---

### **Screenshots or Videos (if applicable)**

image

image

image

---

### **Additional Notes**

- Beni handled the backend implementation of the currency feature.
- Andreas implemented the frontend, added logging, and conducted tests to ensure the feature's compatibility and performance.

### CODE REVIEW

**Comment by dro42** [comment]
image

File: billova_app/static/js/expenses.js [hunk 1]
```diff
@@ -547,25 +610,54 @@ function updateExpense(expenseId, editExpenseForm) {
...
+        }
 
-    if (!expenses.length) {
-        Utils.toggleElementVisibility(SELECTORS.noExpensesCard, true);
-        return;
-    }
+        if (!expenses || !Array.isArray(expenses)) {
+            console.warn("No valid expenses list provided.");
+            Utils.toggleElementVisibility(SELECTORS.noExpensesCard, true);
+            Utils.showNotificationMessage("Error: No expenses data available.", "error");
```
Comment (SergiuCl) [inline billova_app/static/js/expenses.js:627]: Don t show an error to the user if no expenses found, there is already a card displayed which contains the text "No expenses found", no need for additional error message

File: billova_app/static/js/expenses.js [hunk 2]
```diff
@@ -547,25 +610,54 @@ function updateExpense(expenseId, editExpenseForm) {
...
+            throw new Error("List of expenses is not provided or invalid.");
+        }
 
-    expenses.forEach(expense => {
-        addExpenseToTable(expense);
-    });
+        if (!expenses.length) {
+            console.log("No expenses to display.");
+            Utils.toggleElementVisibility(SELECTORS.noExpensesCard, true);
+            Utils.showNotificationMessage("No expenses to display.", "info");
```
Comment (SergiuCl) [inline billova_app/static/js/expenses.js:634]: Don t show an error to the user if no expenses found, there is already a card displayed which contains the text "No expenses found", no need for additional error message

File: billova_app/static/js/expenses.js [hunk 3]
```diff
@@ -547,25 +610,54 @@ function updateExpense(expenseId, editExpenseForm) {
...
+            return;
+        }
+
+        console.log(`Adding ${expenses.length} expenses to the table.`);
+        expenses.forEach(expense => {
+            try {
+                console.log(`Adding expense with ID: ${expense.id}`);
+                addExpenseToTable(expense);
+            } catch (expenseError) {
```
Comment (SergiuCl) [inline billova_app/static/js/expenses.js:643]: This is too much logging in the console. If the user has 1000 expenses, there will be 1000 logs in the console. Please remove the logging

File: billova_app/static/js/expenses.js [hunk 4]
```diff
@@ -547,25 +610,54 @@ function updateExpense(expenseId, editExpenseForm) {
...
+                console.error(`Error adding expense with ID: ${expense.id}`, expenseError);
+                Utils.showNotificationMessage(
```
Comment (SergiuCl) [inline billova_app/static/js/expenses.js:645]: Do not show an error message to the user for each expense which was not added, there could be too many toast messages rendered

File: billova_app/static/js/expenses.js [hunk 5]
```diff
@@ -175,130 +178,189 @@ function setupDomEvents() {
...
-            Utils.showNotificationMessage('Expense added successfully', "success");
+            .then(response => {
+                if (!response.ok) {
+                    console.error("Failed to create expense. Response status:", response.status);
+                    throw new Error(`Failed to create expense: ${response.statusText}`);
+                }
+                return response.json();
+            })
+            .then(data => {
+                console.log("Expense created successfully:", data);
```
Comment (SergiuCl) [inline billova_app/static/js/expenses.js:231]: We don't need to log this in the console. There is already a success message

File: billova_app/static/js/expenses.js [hunk 6]
```diff
@@ -175,130 +178,189 @@ function setupDomEvents() {
...
-        headers: {
-            'Content-Type': 'application/json',
-            'X-CSRFToken': Utils.getCsrfTokenFromForm(createExpenseForm)
-        },
-        body: JSON.stringify(expenseData)
-    })
-        .then(response => {
-            if (!response.ok) {
-                throw new Error('Failed to create expense ' + response.statusText);
+        console.log("Expense data prepared:", expenseData);
```
Comment (SergiuCl) [inline billova_app/static/js/expenses.js:205]: I don't think this is needed to show in the logs

File: billova_app/static/js/expenses.js [hunk 7]
```diff
@@ -547,25 +610,54 @@ function updateExpense(expenseId, editExpenseForm) {
 }
 
 function addExpensesListToTable(expenses) {
-    const expensesTable = document.querySelector(SELECTORS.expenseTable);
-    if (!expensesTable) {
-        Utils.toggleElementVisibility(SELECTORS.noExpensesCard, true);
-        throw new Error("No expenses table found");
-    }
+    try {
+        console.log("Initializing the addition of expenses to the table.");
```
Comment (SergiuCl) [inline billova_app/static/js/expenses.js:614]: I don't think this is needed in the logs

File: billova_app/static/js/expenses.js [hunk 8]
```diff
@@ -602,23 +694,34 @@ function addExpenseToTable(expense, table = document.querySelector(SELECTORS.exp
...
         "Spent": expense.price,
+        "Currency": expense.currency,
         "Note": expense.note,
         "Issuer": expense.invoice_issuer,
         "Category": categories,
         "Actions": actionsContainer.build().outerHTML
     }];
 
     // Use the DataTable's insert method to add the data
+    console.log('Data being added to DataTable:', newData); // <-- Add this
```
Comment (SergiuCl) [inline billova_app/static/js/expenses.js:705]: Not needed, again this function is called from a loop, where you already log each expense (which should also not be logged, see comment above). This will log 1000 logs for 1000 expenses, it is not needed, the data will be visible in the table, we do not need it in logs

File: billova_app/views/expenses.py [hunk 9]
```diff
@@ -1,9 +1,84 @@
...
 
+from billova_app.models import UserSettings
+
+logger = logging.getLogger(__name__)
+
 
 class ExpensesOverview(TemplateView):
     template_name = "Billova/expenses_overview.html"
 
+    def perform_create(self, serializer):
```
Comment (SergiuCl) [inline billova_app/views/expenses.py:14]: I don't understand this part. Why is this actually added here? We used the APIs for the POST request when creating the expense. I think I am missing something

**Review by SergiuCl — CHANGES_REQUESTED** [review]
Please check comments

File: billova_app/views/expenses.py (hunk 9 above)
Comment (dro42) [reply billova_app/views/expenses.py:14]: This function is for pre filing the currency into the form, I tried it with javascript in the expense.js but this did not work
